### PR TITLE
AP_Frsky_Telem: added IMU temperature to FrSky PassThrough telemetry at 0x5001

### PR DESCRIPTION
**Pull request by zenuavsolutions**
Issue #8660
This is my first pr, I have read a lot of code but I'm not sure if this is the right and best way to get the IMU temperature. Since there is no need for quick updates, I simply attached the value to 0x5008 as the 6th parameter. 

There is no IMU heater on my test board, it seems like the temperature is from baro and it is rounded. In yaapus Lua you could now read the temperature as follows:

909: elseif paramId == 6 then
910: IMU_temp = paramValue
911: end

A simple LUA with temperature can be found here:
[https://github.com/zenuavsolutions/Mavlink_Passthrough_Widget](https://github.com/zenuavsolutions/Mavlink_Passthrough_Widget)
img_9206

### CODE REVIEW

File: libraries/AP_Frsky_Telem/AP_Frsky_Telem.cpp [hunk 1]
```diff
@@ -599,6 +600,9 @@ uint32_t AP_Frsky_Telem::calc_param(void)
     case 5:
         param = (uint32_t)roundf(_battery.pack_capacity_mah(1)); // battery pack capacity in mAh
         break;
+    case 6:
+        param = (uint32_t)ins.get_temperature(0); // IMU temperature in degrees C
```
Comment (khancyr) [inline libraries/AP_Frsky_Telem/AP_Frsky_Telem.cpp]: Directly
param = (uint32_t)AP::ins().get_temperature(0);

**Comment by khancyr** [comment]
Hello,
Thanks for your contribution. 
Could you edit your commit message to AP_Frsky_Telem: add IMU temperature to 0x5008 (see http://ardupilot.org/dev/docs/submitting-patches-back-to-master.html). Thanks

**Comment by zenuavsolutions** [comment]
@floaledm 
I tested this with the following boards: Pixhawk 1 non 3DR/clone, Pixhawk 2.1, PixRacer. The temperature is measured as i know in the barometer. For Pixhawk 2 users the temperature is important as it is recommended to heat up IMU to 60 ° C.

I was in contact with the guys from Craft & Theory and made this suggestion, they think the protocol would be very "full" and every byte valuable, so I did not create a new ID although it would be interesting for Pixhawk 2 users as well set target temperature to see. 

The update speed of the temperature does not need to be fast, so I chose 5008 as a 6th value

**Comment by floaledm** [comment]
@zenuavsolutions I'm the guy from C&T who talked to you before. The main concern is always to minimize bandwidth usage because there is so few packets that can be transmitted per second.

I think we should see if it wouldn't be too complicated to send the data only for the boards for which the information is useful, so as not to waste bandwidth on the other boards AND see what is the minimally acceptable update rate and precision so as to minimize bandwidth usage for the board for which that info is useful.

So to further explain, adding the info to ID 0x5001 would not affect performance at all (since that packet is sent anyway). Adding as a 0x5008 does affect the update rate of the rest of the parameter data (since one parameter value is sent per second). Adding as its own sensor ID has the benefit of allowing you to strictly control the update rate, but you do use 32 bits for the info, which I think is wasteful, considering that the temp could easily be transmitted on 8 bits or even less.

So, I think we should simply first establish a) what is an acceptable update rate and b) what is the least amount of bits we can use to meaningfully represent the temperature, given what we can envision is is most useful for (e.g., verifying that the IMU has heated up to 60C or higher).

For example, if you're only concerned about "Did the IMU heat up to above 60C?", yes or no, then one bit is sufficient. If you are also wanting to see how fast the temp is rising, then more bits are necessary.

I'm not trying to be stingy here, but we should be conscientious of bandwidth usage in general on the FrSky link... Please let me know what you think is a reasonable update rate and way to represent the temp (or to represent whether the temp is above 60C or not)...

**Comment by zenuavsolutions** [comment]
@floaledm 
only one bit does not make sense. I've got a Pixhawk 2 that never reaches 60° but only 58.8. Also, you do not know then how much wait time is still to be expected. Therefore, I would like to transfer the temperature. The IMU temperature is also useful for hardware without heating, e.g. in very cold and very hot regions or seasons

The other values in 0x5008 are actually static (mavtype, failsafe voltage, batt capacity) so I do not think that it interferes with the temperature here. 0.5 hertz is enough update rate.

For 0x5001 I would have another idea about which helicopter pilots would be very happy: RPM, but this is off topic ;-)

**Comment by floaledm** [comment]
@zenuavsolutions I understand what you'd like to do and I'm sorry to insist but I think we would all benefit if a few questions were answered to precisely address the need here because there are only about 32 4-byte packets that can be transmitted per second,and that's for all equipment connected to the SPort bus, so every bit is valuable. One bit does make sense if all you need is the answer to a yes or no question (60°C was an example, but it could be another yes or no question that you'd try to answer - it simply depends on the need). As you indicate, more bits are probably required to see the temperature progression, so then let's figure out just how many and at what rate... It's that simple.

Plugging it into 0x5008 works, sure, but it's going to be at an update rate lower than 0.5Hz (the protocol cycles through one parameter per second) since there are 6 "cases" right now, and if there are more parameters added, that update rate will become even lower. Additionally, there is no good reason in my view to use an entire packet (32 bits) for one temperature value. Otherwise, there's protocol 4 which is not concerned with bandwidth management that you can work with... I hope you understand that, while it may not matter for the trivial addition you want to make, we need to do this correctly the first time around because a) there is always other things to add (example in point, RPM), b) multiple developers depend on this protocol (and therefore its backwards compatibility) for their systems, and c) unfortunately the bandwidth is a strict limitation due to the hardware. If we don't put much thoughts into bandwidth management, the protocol will end up not being that useable and upgradable...

Again, we need to start with these simple questions (I have my own answers but this is a discussion, so I'd like to hear your's and other's):
1. do all boards provide IMU temperature, and for those that don't have a heated IMU, is that temperature representative of the hardware or environmental temperature in general?
2. what is the minimally acceptable update rate for the IMU temperature?
3. what is the minimally acceptable bit size representation for the IMU temperature?
4. are there other temperature we would like to send, or other info that is somewhat related, that we could send all together in a new packet ID; e.g., ESC temperatures, similar in principle to the FrSky 80A ESCs (https://www.frsky-rc.com/wp-content/uploads/Downloads/Manual/FrESC32_80A/Manual%20-%20FrESC32_80A.pdf)?

**Comment by zenuavsolutions** [comment]
Unfortunately, I do not have an answer to all questions. I can only say that the majority, or better, all users / developers with whom I have contact, consider communicating the IMU temperature sensible and important, including Philip Rowse from ProfiCNC.

Question 1: I can not answer as I have only 3 different boards and tested. These have all transmitted a temperature, (even that cheap Pixhawk 1 clone). But I can interview hardware developers.

To question 2: for my opinion even 0.1 Hz would be enough, given the waiting time of about 20-30 minutes to 60° are reached. It's not a critical value.

to question 3: That depends also on question 4. In view of the fact that largely static values ​​are repeatedly transmitted in 5008, I can not quite understand the objection but 8 bits would certainly suffice, so further values ​​could be transmitted in 5008 case 6 at the same time. On the other side, i don't know the possible range of temperature values in Ardupilot code...

to question 4: would it not make more sense for hardware minorities or special cases to allow something like a userspace. Suppose someone flies with a turbine, RPM and exhaust temperature are important, others wants to see the ESC or IMU temperatures... i don't know much about DShot Ardupilot telemetry capabilities, i used this only on Betaflight racecopters until know. I think this is also not related to all users, as not all have IMU heaters.

maybe it would make sense to make the questions accessible to a wider audience?

**Comment by floaledm** [comment]
@zenuavsolutions Thanks for taking the time to answer in detail. Without complicating things too much, I definitely welcome the input from others so that we can come up with a good solution,. At the same time, I don't want to stall this and would like to get it approved as soon as reasonably possible.

Question 1: It would have been nice to have some breakdown on this, but I think we can move on with the PR regardless of the answer and adjust later if need be, using the info you may find.

Question 2: I agree with your assessed 0.1Hz minimum update rate. Yes, it's subjective, but so is one's appreciation regarding the rate at which the temp rises, so that should work just fine.

Question 3: You're right that the params ID is mean for static and user modified values that most likely don't change at all in a given flight, and that ideally we would want them to be transmitted something like just once at init, but since there's currently no mechanism to request the params to be transmitted and to check if the params are correctly received (ACK/NAK?), this is the next best mechanism that was found (just send them in cycle periodically at a reasonably low update rate as to no affect the rest of the bandwidth too much). If you have suggestions to make the transmission of the param values more efficient, that would be much appreciated (see question 4 comments below, including why 0x5008 should maybe be strictly reserved for parameters and static values).

8 bits is a good start, but establishing what is an acceptable minimum would help us figure out where we can place it. It's the same process that's been done with all the info in the protocol, which ultimately allows the protocol to transfer a fairly large amount of information in a relatively small number of packets. To find that minimum number of bits without scratching out heads too much, we should, in my view, establish the min and max values that we need to see, and a resolution. For example, you could determine that you need to see the progression of the temp from 20C up to 80C, with a resolution of 1C, and establish that values below 20C are too cold, and that values above 80C are too hot, in which case 6 bits would be sufficient. A single value (lets say the one which would otherwise correspond to 19C) could then be used to report "below 20C", and similarly "81" could be used to report "above 80C/too hot!" Let me know what your thoughts are on the range and resolution needed.

Question 4: I agree with you there as well since the info to be transmitted will depend on the type of vehicle/equipment and the need. Really what needs to be coded in ArduPilot is a two-way link since the hardware and OpenTX are capable (oops, just opened another can of worms!). You could envision starting with the protocol as it stands with its default packets and update rates at init, but then a device could communicate with the flight controller and ask for different things to be transmitted at different rates, as needed (in a sense, similar to MAVLink data streams: http://mavlink.org/messages/common#DATA_STREAM), including requesting all the params when the device connects, just like ground stations do via MAVLink. By the way, this is why I'm reluctant to put a value that fluctuates (the IMU temp) in with the rest of the params, because ultimately, I would like to see a mechanism where all the params are requested at init, whereas the IMU temp would be best polled or sent periodically... Even if this two-way mechanism gets implemented though, I think that we'd still need to have this protocol with defaults in place for backwards compatibility. Anyway, that is also thoughts for another day...

At this point, I lean towards a) creating a new dataID for either "IMU information" or "temperature information" with a 0.1Hz rate, or b) adding the temp data to 0x5001 (which right now has 20 bits left), especially if we can get it down to 6 bits or less...

File: libraries/AP_Frsky_Telem/AP_Frsky_Telem.cpp (hunk 1 above)
Comment (zenuavsolutions) [reply libraries/AP_Frsky_Telem/AP_Frsky_Telem.cpp]: I made a commit to chanche this

**Comment by zenuavsolutions** [comment]
I've written with Philip Rowse, he also thinks that a range of 20-80 degrees or 6-8 bits should be enough. So what can we do next? :)

**Comment by floaledm** [comment]
Since we have decided on a minimum update rate (0.1Hz+), we just need to agree on:
- the range (no objections so far on 20-80C range, but I did pick that range kind of arbitrarily to illustrate what I was explaining) and,
- the precision (is 1C ok? can we go for less precise - e.g., 2C precision?)

Then, we can edit this PR to plug it where it will interfere the least with the rest of the data stream.

Ultimately, if 6 or less bits is possible for a meaningful representation of the IMU temp, I lean towards plugging it in 0x5001 (even if it has a higher update rate), otherwise, I think it should be in a new data ID which we should possibly structure so that we can expand it later (as I said, structure it for "IMU information" or "temperature information").. whatever we decide may be most pertinent.

**Comment by floaledm** [comment]
@zenuavsolutions I did not mean to stall this... Once we have the answers to the questions above, we can make the necessary changes to this PR.

**Comment by zenuavsolutions** [comment]
I think 2 degree resolution would certainly be sufficient, but I think it is a bit too pragmatic and would prefer 1 degree. The range of 20-80 is ok in my opinionI think 2 degree resolution would certainly be sufficient, but I think it is a bit too pragmatic and would prefer 1 degree. The range of 20-80 is ok in my opinion

**Comment by zenuavsolutions** [comment]
I have no preferences and think you are more competent to decide, for me ID 5001 is also ok.

**Comment by floaledm** [comment]
Sounds good. Can you revise this PR by placing the IMU temp at maybe the last 6 bits of 0x5001? Thanks.

**Comment by zenuavsolutions** [comment]
@floaledm I modified the code as suggested by you. Since I am autodidact I do not know if I have generated the 6 bit value correctly. There are probably more elegant ways ;-)

**Comment by floaledm** [comment]
Thanks. Let me know when it's ready for review and I'll take a look.

**Comment by zenuavsolutions** [comment]
@floaledm checks have passed again! I think I'm ready!
also made some successful tests with different boards and modified LUA Widget:
https://github.com/zenuavsolutions/Mavlink_Passthrough_Widget

img_9331
img_9332

File: libraries/AP_Frsky_Telem/AP_Frsky_Telem.cpp [hunk 2]
```diff
@@ -682,6 +682,13 @@ uint32_t AP_Frsky_Telem::calc_batt(uint8_t instance)
 uint32_t AP_Frsky_Telem::calc_ap_status(void)
 {
     uint32_t ap_status;
+    
+    // prepare imu temp to a 6 bit value with a range from 20° - 83° value offset is 19
+    // 1 = 20°, 63 = 82°
```
Comment (floaledm) [inline libraries/AP_Frsky_Telem/AP_Frsky_Telem.cpp]: You may want to clarify the comments here, whereas, for the value stored on the 6 MSBs of ap_status: 0 (0x00) represents 19°C or less, and 63 (0x3f) represents 82°C or more (there is no value representing 83°C, or even 82°C).
1 (0x01) = 20°C, 62 (0x3e) = 81°C (63 not equal to 82°C, but instead means a value of 82°C or more).

File: libraries/AP_Frsky_Telem/AP_Frsky_Telem.cpp [hunk 3]
```diff
@@ -682,6 +682,13 @@ uint32_t AP_Frsky_Telem::calc_batt(uint8_t instance)
...
+    uint32_t imutemp = AP::ins().get_temperature(0) - 19;
+    if (imutemp <= 0) imutemp = 0;
+    if (imutemp >= 63) imutemp = 63;
```
Comment (floaledm) [inline libraries/AP_Frsky_Telem/AP_Frsky_Telem.cpp]: Have you tested your conversion routine alone in a compiler? Have you tested with temperatures of 18°C or less?

Using a standalone compiler to test your conversions, I found the following:
- A "get_temperature" of 18°C or less will end up giving a value of 63 to imutemp...
- Also there's an issue with rounding of the float returned by get_temperature(): for example, a "get_temperature" of 19.9°C will end up giving a value of 0 to imutemp (when it should really be rounded to 1), and as another example a "get_temperature" of 81.5°C will end up giving a value of 62 to imutemp (when it should really be rounded to 63).

There appears to be no need to create a variable imutemp (anything helps I guess to optimize memory). All can fit on one line (even if it may not look quite as readable). I propose adding to AP_Frsky_Telem.h the following:
`#define AP_IMU_TEMP_ORIGIN          19`
and replacing your code with this one line:
`ap_status =| ((AP::ins().get_temperature(0) < AP_IMU_TEMP_ORIGIN) ? 0 : (AP::ins().get_temperature(0) > (AP_IMU_TEMP_ORIGIN + AP_IMU_TEMP_LIMIT)) ? AP_IMU_TEMP_LIMIT : (uint8_t)roundf(AP::ins().get_temperature(0) - AP_IMU_TEMP_ORIGIN))<<AP_IMU_TEMP_OFFSET;`
I tested this routine in a standalone compiler with "get_temperatures" values such as 18.5, 19, 19.4, 19.5, 20, 81.4, 81.5, 82, etc. and it all seems to round up nicely... Let me know what you think.

File: libraries/AP_Frsky_Telem/AP_Frsky_Telem.h [hunk 4]
```diff
@@ -95,6 +96,8 @@ for FrSky SPort Passthrough
 #define AP_ARMED_OFFSET             8
 #define AP_BATT_FS_OFFSET           9
 #define AP_EKF_FS_OFFSET            10
+#define AP_IMU_TEMP_LIMIT           0x3F
+#define AP_IMU_TEMP_OFFSET          25
```
Comment (floaledm) [inline libraries/AP_Frsky_Telem/AP_Frsky_Telem.h]: Shouldn't AP_IMU_TEMP_OFFSET have a value of 26, if the corresponding value is populated in the 6 MSBs of 0x5001?
For example, compare to HOME_BEARING_OFFSET which has a value of 25, and the corresponding value is coded on the last 7 MSBs of 0x5004.

Comment (floaledm) [reply libraries/AP_Frsky_Telem/AP_Frsky_Telem.h]: As discussed above, I would add this definition:
#define AP_IMU_TEMP_ORIGIN 19

**Comment by floaledm** [comment]
@zenuavsolutions I provided my comments. Thanks.

**Comment by zenuavsolutions** [comment]
@floaledm thank you very much for help! So this pr is now ready?

**Comment by zenuavsolutions** [comment]
@floaledm I have implemented your suggestions for improvement in the last commit and would like to thank you for your great help. It's people like you that make Ardupilot better and better!

**Comment by floaledm** [comment]
PR looks good to me now. I can't push myself, so if anybody out there is willing to do it, that would be much appreciated!

**Comment by zenuavsolutions** [comment]
Is there anybody out there who can push/merge this commit? Would be great to have this feature in Copter 3.6 :-)

**Comment by lvale** [comment]
If we would want to send temperature on the telemetry link, I think would be more relevant to send motor temperatures for non electric vehicles, than sending the IMU's. As it is, it's only a low informational value to one single FC model, and has no possible way to be acted upon without a GCS, so it is only important to be displayed on a regular GCS.

**Comment by zenuavsolutions** [comment]
@lvale I do not agree with you. The temperature is determined on all boards with barometer (Pixhawk 1 / 2, Pixracer etc.), not only in IMUs with heater. And it is extremely annoying to be able to see the IMU temperature only if a laptop with GCS is available. I also own a turbine helicopter, but the ECU of the turbine has its own telemetry, it does not make any sense for me to transmit the exhaust temperature via the flight control...

**Comment by proficnc** [comment]
@lvale this has nothing to do with motor temperatures. This is specifically about IMU temperature.

File: libraries/AP_Frsky_Telem/AP_Frsky_Telem.cpp [hunk 5]
```diff
@@ -695,6 +695,8 @@ uint32_t AP_Frsky_Telem::calc_ap_status(void)
     ap_status |= (uint8_t)(AP_Notify::flags.failsafe_battery)<<AP_BATT_FS_OFFSET;
     // bad ekf flag
     ap_status |= (uint8_t)(AP_Notify::flags.ekf_bad)<<AP_EKF_FS_OFFSET;
+    // IMU temperature: offset -19, 0 means temp =< 19°, 63 means temp => 82°
+    ap_status |= ((AP::ins().get_temperature(0) < AP_IMU_TEMP_ORIGIN) ? 0 : (AP::ins().get_temperature(0) > (AP_IMU_TEMP_ORIGIN + AP_IMU_TEMP_LIMIT)) ? AP_IMU_TEMP_LIMIT : (uint8_t)roundf(AP::ins().get_temperature(0) - AP_IMU_TEMP_ORIGIN))<<AP_IMU_TEMP_OFFSET;
```
Comment (tridge) [inline libraries/AP_Frsky_Telem/AP_Frsky_Telem.cpp]: this line need simplification, it is very hard to follow what it does. Can you please break this up into multiple lines?

Comment (tridge) [reply libraries/AP_Frsky_Telem/AP_Frsky_Telem.cpp]: also, don't call get_temperature() 3 times

Comment (zenuavsolutions) [reply libraries/AP_Frsky_Telem/AP_Frsky_Telem.cpp]: your second request is done and committed

File: libraries/AP_Frsky_Telem/AP_Frsky_Telem.cpp [hunk 6]
```diff
@@ -695,6 +696,8 @@ uint32_t AP_Frsky_Telem::calc_ap_status(void)
     ap_status |= (uint8_t)(AP_Notify::flags.failsafe_battery)<<AP_BATT_FS_OFFSET;
     // bad ekf flag
     ap_status |= (uint8_t)(AP_Notify::flags.ekf_bad)<<AP_EKF_FS_OFFSET;
+    // IMU temperature: offset -19, 0 means temp =< 19°, 63 means temp => 82°
+    ap_status |= ((imu_temp < AP_IMU_TEMP_ORIGIN) ? 0 : (imu_temp > (AP_IMU_TEMP_ORIGIN + AP_IMU_TEMP_LIMIT)) ? AP_IMU_TEMP_LIMIT : (uint8_t)roundf(imu_temp - AP_IMU_TEMP_ORIGIN))<<AP_IMU_TEMP_OFFSET;
```
Comment (tridge) [inline libraries/AP_Frsky_Telem/AP_Frsky_Telem.cpp]: please do this as multiple if statements to make it more readable

**Comment by magicrub** [comment]
there are more commits than Lines-of-Code changed.. Needs a squash into a single commit. If you need help with that, let us know. We can do it for you.

**Comment by floaledm** [comment]
Thanks for your comments and for your offer to help Jochen get this through. Hopefully we're almost there!

**Comment by zenuavsolutions** [comment]
@magicrub Yes, I could use help, thank you! The number of commits came about as a result of floaledm and I discussed the best option.

**Comment by SergeyBokhantsev** [comment]
Boards that have no IMU temp will always send 0 via telemetry, so it will be interpret as +19 degrees at telemetry client, that is not very good.
Should 0 explicitly mean no temperature?

**Comment by zenuavsolutions** [comment]
@floaledm my skills are lower then yours, could you show me how to code the comment from tridge? Many people would see this feature in 3.6 :-)

File: libraries/AP_Frsky_Telem/AP_Frsky_Telem.cpp (hunk 6 above)
Comment (floaledm) [reply libraries/AP_Frsky_Telem/AP_Frsky_Telem.cpp]: This reads as:
if (imu_temp < AP_IMU_TEMP_ORIGIN)
    ap_status |= 0;
else if (imu_temp > (AP_IMU_TEMP_ORIGIN + AP_IMU_TEMP_LIMIT))
    ap_status |= AP_IMU_TEMP_LIMIT;
else ap_status |= (uint8_t)roundf(imu_temp - AP_IMU_TEMP_ORIGIN)<<AP_IMU_TEMP_OFFSET;

This may not be 100% correct so please check...

**Comment by floaledm** [comment]
@SergeyBokhantsev I think you have a good point, at the same time, since the temp function returns 0 if there is no temperature sensor, that also corresponds to a valid temperature value (0°C). Ultimately, is it sufficient to know that there is either no temp or that the temp is 19°C or less (current implementation) or is it really important to know for sure when there is no temp reported (by distinguishing as much as possible from actual temp values)?

@zenuavsolutions Sure thing (see my inline comment next to Tridge's). I used a couple of these: https://en.wikipedia.org/wiki/%3F:#C
You also still need to squash all into one commit... I like to use the text editor based method of `git rebase -i HEAD~x` where x is the number of commits you want to look at and modify.

**Comment by SergeyBokhantsev** [comment]
>> or is it really important to know for sure when there is no temp reported (by distinguishing as much as possible from actual temp values)?

No, I believe it could be as is. It just a matter of design agreement.
So it is up to client side to know if IMU Temp exist or not. So client could interpret zero as 19C or as NO DATA.

File: libraries/AP_Frsky_Telem/AP_Frsky_Telem.cpp (hunk 6 above)
Comment (zenuavsolutions) [reply libraries/AP_Frsky_Telem/AP_Frsky_Telem.cpp]: thank you @floaledm i have read your code simplification to late, have done nearly the same way...

**Comment by zenuavsolutions** [comment]
@magicrub I tried to squash my commits in Sourcetree, unfortunately without success, could you please help?

**Comment by zenuavsolutions** [comment]
@OXINARF the requested code simplification is done. But i can‘t squash my commits, could somebody help?

**Comment by zenuavsolutions** [comment]
see here for my new clean pull request #9464

**Comment by zenuavsolutions** [comment]
@magigrub @tridge
I made a cleaner pullrequest: #9464
Could you push it into master please?
